### PR TITLE
Allow CombinerGenerator to take a file of positions. 

### LINKED ISSUE #16834: Allow CombinerGenerator to get mesh positions from a file
## Reason
In Cardinal, we are creating BISON meshes for 100,000+ pebble elements by translating the same Exodus mesh to the centroid of each pebble, which results in enormous `positions` sections for the `CombinerGenerator`. We already generate these positions in a text file format, so it would be convenient if we could simply grab those positions from the file rather than copying-and-pasting thousands of lines all the time.

## Design
Similar to the `positions`/`positions_file` arguments that can be set for MultiApps, we can add a `positions_file` argument to the `CombinerGenerator` that should read the mesh positions from a file.

## Impact
New feature, very convenient for problems with heavy usage of repeated geometries in `CombinerGenerator`

### PR DESCRIPTION
**Pull request by aprilnovak**
Adds a `positions_file` option to `CombinerGenerator` to make specifying huge numbers of positions easier. I followed the same approach used in `MultiApp`, including the parallel communication. Feedback on the validity of that basically copy-paste file reading part would be great!

Closes #16834

### CODE REVIEW

**Comment by loganharbour** [comment]
Would you object to moving most of the duplicated code into a utility instead that could be called? I think we'd definitely prefer that over a ton of copy-pasta.

**Comment by aprilnovak** [comment]
Not at all! That's definitely preferred - will add tomorrow.

**Comment by loganharbour** [comment]
> Not at all! That's definitely preferred - will add tomorrow.

👍🏼 great, thanks! Its use here is definitely appropriate and I think this is good otherwise.

File: framework/include/meshgenerators/CombinerGenerator.h [hunk 1]
```diff
@@ -40,9 +43,15 @@ class CombinerGenerator : public MeshGenerator
...
 
-  /// The (offsets) positions for each mesh
+  /// The (offsets) positions for each mesh if specified as a vector of points
   const std::vector<Point> & _positions;
 
+  /// The (offsets) positions for each mesh if specified as vector of points in file
+  const std::vector<FileName> & _positions_file;
+
+  /// The (offset) positions for each mesh
+  std::vector<Point> _generator_positions;
```
Comment (loganharbour) [inline framework/include/meshgenerators/CombinerGenerator.h:53]: For simplicity, I say just keep the single `_positions` variable. If you want to keep it const, you could make `fillPositions()` just return a vector depending on the parameter combo.

File: framework/include/meshgenerators/CombinerGenerator.h [hunk 2]
```diff
@@ -31,6 +31,9 @@ class CombinerGenerator : public MeshGenerator
 
   std::unique_ptr<MeshBase> generate() override;
 
+  /// Fill the offset positions for each mesh
+  virtual void fillPositions();
```
Comment (loganharbour) [inline framework/include/meshgenerators/CombinerGenerator.h:35]: Doesn't really have a need to be virtual

File: framework/include/meshgenerators/CombinerGenerator.h (hunk 1 above)
Comment (aprilnovak) [reply framework/include/meshgenerators/CombinerGenerator.h:53]: hmm, but if `getParam` (if user does `positions = '1 1 1'`) requires `const`, and `push_back` (if user does `positions_file = 'positions.txt'`) won't allow `const`, is that possible?

Comment (aprilnovak) [reply framework/include/meshgenerators/CombinerGenerator.h:53]: Nevermind, ignore my brain lapse here...

**Comment by aprilnovak** [comment]
Not sure what would be best to do about the debug failure - I added a `mooseAssert` to check that the data is the correct size when reading by rows (I was just mirroring the assert statement in `DelimitedFileReader` for the column-based format), because I noticed that this check was missing.

But the failing test looks to be caused by my assert now terminating the program, as opposed to a specially-written `mooseError` elsewhere.

Basically, the error is still caught, but perhaps with a slightly less user-friendly message (but I feel that the assert statement is better placed where it is because it checks any functionality that uses the `DelimitedFileReader` util). I'll default to your preference here - (A) remove my assert statement, (B) keep the assert and change this test.

My assert statement:
```
misc/check_error.function_file_test3: Assertion `raw.size() % n_cols == 0' failed
misc/check_error.function_file_test3: The raw data is not evenly divisible by the number of rows.
misc/check_error.function_file_test3: at /opt/civet/build_0/moose/framework/src/utils/DelimitedFileReader.C, line 101
misc/check_error.function_file_test3: Stack frames: 27
misc/check_error.function_file_test3: 0: libMesh::print_trace(std::ostream&)
misc/check_error.function_file_test3: 1: MooseUtils::DelimitedFileReader::read()
```

**Comment by aprilnovak** [comment]
Ok I ended up just removing my assert statement to avoid that test error. The `CombinerGenerator` will still catch the error that was caught by the assert anyways.

Should be good to go now!

**Comment by aprilnovak** [comment]
hey @loganharbour, were there any other changes you'd like to see in this PR?

**Comment by loganharbour** [comment]
Sorry for the delay. Will get to this tonight or tomorrow morning.

**Comment by aprilnovak** [comment]
No worries - thanks for reviewing!

**Review by loganharbour — APPROVED** [review]
Alright - sorry about that.

My original intention was to reduce code further by adding a tool that takes the object and the parameter name, and does all the error checking within there and can actually call paramError on the object. However... after thinking about that more, and also adding support for a vector of names, etc.... I realized it was too much trouble.

**Comment by aprilnovak** [comment]
> a tool that takes the object and the parameter name, and does all the error checking

That would be really nice, actually! Because there's quite a few places in Cardinal where we're doing this exact same file reading process. The only reason I made this implementation an `std::vector<FileName>` was because that was how it was implemented in the `MultiApp`s (does anyone actually use a vector of filenames there? I don't know).

If it would help simplify your utility and you don't forsee any issue with getting rid of the option to set `positions_file = 'file1.txt file2.txt'` I can update this PR.

**Comment by aprilnovak** [comment]
Hey just checking in - do you have any more thoughts on this PR?

**Comment by loganharbour** [comment]
If you have the idea or the time to think about a way to clean up the file input in a utility used by MooseObjects (takes parameters, a param name, returns a vector of points), please let me know :)

Good for this to go in after re-running tests real quick.

**Pull request merged**